### PR TITLE
Adds code formatting check in the CI

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
ref: https://github.com/ada-url/ada/issues/170 and https://github.com/ada-url/ada/pull/243#issuecomment-1478748447
Hello There! @lemire @anonrig 
How do we proceed with the format of the whole codebase? Is there any big (code)change currently in progress for this repo? If not, can we run the format for all files?

### CODE REVIEW

**Comment by anonrig** [comment]
I think you can format it in this pull request with an additional commit.

File: .github/workflows/lint_and_format_check.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+permissions:
+  contents: read
+
+jobs:
+  ubuntu-build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3 
+      - name: Install ICU
+        run: sudo apt-get install -y libicu-dev pkg-config
```
Comment (anonrig) [inline .github/workflows/lint_and_format_check.yml:20]: We don't need ICU in this repository :-)

File: .github/workflows/lint_and_format_check.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+      - name: Install clang-format
+        run: |
+          sudo apt update && sudo apt install clang-format-15 -y
```
Comment (anonrig) [inline .github/workflows/lint_and_format_check.yml:24]: We are using `clang-format-15` but `clang++-14` as `CXX` is this expected? cc @lemire

File: scripts/lint_and_format.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import os
+import sys
+import subprocess
+import argparse
+
+parser = argparse.ArgumentParser(
+    description="Format C/C++ code using clang-format."
```
Comment (anonrig) [inline scripts/lint_and_format.py:7]: Ideally we should also add `ruff` as a linter for all of our Python scripts. It will be a good (additional) PR

**Review by anonrig — APPROVED** [review]

**Comment by anonrig** [comment]
Should we remove the existing hooks? It doesn't seem to be necessary anymore.

File: .github/workflows/lint_and_format_check.yml (hunk 2 above)
Comment (miguelteixeiraa) [reply .github/workflows/lint_and_format_check.yml:24]: clang-format is a standalone tool, so I don't think it needs to be tied to the actual version of clang++.
But sure there are differences in the way these major versions (14, 15) format the code and also in the "language features"/things they are able to format.

**Comment by miguelteixeiraa** [comment]
I just crashed everything yayyyy  :sweat_smile:
I'll figure it out when I have a little more time

**Comment by lemire** [comment]
If you reorder the include files, you may break the amalgamation. I recommend you do not reorder the include files.

**Comment by miguelteixeiraa** [comment]
> If you reorder the include files, you may break the amalgamation. I recommend you do not reorder the include files.

Also, we would have problems with the includes order of idna.cpp, normalization.cpp, and probably more. Just added a `SortIncludes: false` to our .clang-format conf

**Comment by miguelteixeiraa** [comment]
> > If you reorder the include files, you may break the amalgamation. I recommend you do not reorder the include files.
> 
> Also, we would have problems with the includes order of idna.cpp, normalization.cpp, and probably more. Just added a `SortIncludes: false` to our .clang-format conf (probably have to also add it in ada)

Actually I think they are all about the amalgamation like @lemire said. I can try to refactor the amalgamation script + related things to not depend on the import order if the current solution is not okay

**Comment by lemire** [comment]
There are still issues even if you preserve the original include order?

**Comment by miguelteixeiraa** [comment]
> There are still issues even if you preserve the original include order?

No issues when preserve the original includes order (don't know if it is in the original order now, but I can make it)

**Comment by lemire** [comment]
The exact order does not matter, but some orders will break the amalgamation script.

**Comment by miguelteixeiraa** [comment]
> It seems the ordering issue is resolved.

That's correct.

**Pull request merged**